src/stronghold/mod.rs

```diff
@@ -209,25 +222,26 @@ impl StrongholdAdapter {
     pub async fn set_password(&mut self, password: &str) -> Result<()> {
         // In a closure so there is no deadlock when calling `self.read_stronghold_snapshot()`
         {
-            let mut key = self.key.lock().await;
-            if key.is_some() {
+            let mut key_provider = self.key_provider.lock().await;
+            if key_provider.is_some() {
                 return Err(crate::Error::StrongholdPasswordAlreadySet);
             }
-            *key = Some(self::common::derive_key_from_password(password));
+            *key_provider = Some(KeyProvider::try_from(
+                (*self::common::derive_key_from_password(password)).clone(),
+            )?);
         }
 
         // Try to read a snapshot to check if the password is correct
-        if self.snapshot_path.is_some() {
-            let result = self.read_stronghold_snapshot().await;
-            if let Err(err) = result {
-                // TODO: replace with actual error matching when updated to the new Stronghold version
-                if let crate::Error::StrongholdProcedureError(ref err_msg) = err {
-                    if !err_msg.contains("IOError") {
-                        // If loading the snapshot failed but wasn't an IOError, then the password was incorrect and we
-                        // clear it
-                        *self.key.lock().await = None;
-                        return Err(err);
-                    }
+        let result = self.read_stronghold_snapshot().await;
+        if let Err(err) = result {
+            // TODO: replace with actual error matching when updated to the new Stronghold version
+            // TODO: but the error is not an enum
+            if let crate::Error::StrongholdProcedureError(ref err_msg) = err {
+                if !err_msg.to_string().contains("IOError") {
```

Comment:
Shouldn't we get a `SnapshotFileMissing` error now?

Reply:
https://github.com/iotaledger/stronghold.rs/commit/27c20730f9ab7ab6ab3807bdc1fac31e39d80dc0

Reply:
Yes, I changed the one from `try_build` but forgot this one, will change